ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts

```diff
@@ -0,0 +1,137 @@
+    sogsMutationCache.push(entry);
+    window.log.info('SOGS Mutation Cache: Entry added!', entry);
+  }
+}
+
+export function updateMutationCache(entry: SogsV3Mutation, seqno: number) {
+  if (!verifyEntry(entry)) {
+    window.log.error('SOGS Mutation Cache: Entry verification on update failed!');
+  } else {
+    const entryIndex = findIndex(sogsMutationCache, entry);
+    if (entryIndex >= 0) {
+      const updatedEntry = entry;
+      updatedEntry.seqno = seqno;
+      sogsMutationCache[entryIndex] = updatedEntry;
+      window.log.info('SOGS Mutation Cache: Entry updated!', updatedEntry);
+    } else {
+      window.log.error('SOGS Mutation Cache: Updated failed! Cannot find entry', entry);
+    }
+  }
+}
+
+export async function processMessagesUsingCache(
+  server: string,
+  room: string,
+  message: OpenGroupMessageV4
+) {
+  const updatedReactions = message.reactions;
+  const roomMatches: Array<SogsV3Mutation> = filter(sogsMutationCache, { server, room });
+
+  if (roomMatches?.length) {
+    for (const roomMatch of roomMatches) {
+      if (message.seqno && roomMatch.seqno && roomMatch.seqno <= message.seqno) {
+        const removedEntry = remove(sogsMutationCache, roomMatch);
+        window.log.info('SOGS Mutation Cache: Entry ignored and removed!', removedEntry);
+      } else if (
+        !message.seqno ||
+        (message.seqno && roomMatch.seqno && roomMatch.seqno > message.seqno)
+      ) {
+        for (const reaction of Object.keys(message.reactions)) {
+          const reactionMatches = filter(sogsMutationCache, {
+            server,
+            room,
+            changeType: ChangeType.REACTIONS,
+            metadata: {
+              messageId: message.id,
+              emoji: reaction,
+            },
+          });
+          if (reactionMatches?.length) {
+            for (const reactionMatch of reactionMatches) {
+              switch (reactionMatch.metadata.action) {
+                case 'ADD':
+                  updatedReactions[reaction].you = true;
+                  updatedReactions[reaction].count += 1;
+                  window.log.info(
+                    'SOGS Mutation Cache: Added our reaction based on the cache',
+                    updatedReactions[reaction]
+                  );
+                  break;
+                case 'REMOVE':
+                  updatedReactions[reaction].you = false;
+                  updatedReactions[reaction].count -= 1;
+                  window.log.info(
+                    'SOGS Mutation Cache: Removed our reaction based on the cache',
+                    updatedReactions[reaction]
+                  );
+                  break;
+                default:
+                  window.log.warn(
+                    'SOGS Mutation Cache: Unsupported metadata action in OpenGroupMessageV4',
+                    reactionMatch
+                  );
+              }
+            }
+
+            const removedMatches = remove(sogsMutationCache, ...roomMatches);
```

Comment:
I am not too sure if this is the right spot for this. Shouldn't you do this outside of the main `for..of`

Reply:
or you didn't take into account that `remove` mutates the array?

Reply:
Good catch, I had put it in the wrong location by mistake. I'm aware it mutates the array 👍